relay-server/src/endpoints/common.rs

```diff
@@ -344,85 +299,65 @@ where
     let project_key = meta.public_key();
     let start_time = meta.start_time();
     let config = request.state().config();
-
-    let envelope_context = Rc::new(RefCell::new(EnvelopeContext::from_request(&meta)));
+    let event_id = Rc::new(RefCell::new(None));
 
     let future = extract_envelope(&request, meta)
         .into_future()
-        .and_then(clone!(config, envelope_context, |mut envelope| {
-            envelope_context.borrow_mut().update(&envelope);
+        .and_then(clone!(config, event_id, |mut envelope| {
+            *event_id.borrow_mut() = envelope.event_id();
 
             // If configured, remove unknown items at the very beginning. If the envelope is
             // empty, we fail the request with a special control flow error to skip checks and
             // queueing, that still results in a `200 OK` response.
             utils::remove_unknown_items(&config, &mut envelope);
 
+            let mut envelope_context = EnvelopeContext::from_envelope(&envelope);
             if envelope.is_empty() {
-                // envelope is empty, cannot send outcomes
+                envelope_context.reject(Outcome::Invalid(DiscardReason::EmptyEnvelope));
```

Comment:
Most of the rejections have been pushed deep into message handlers, this one being a notable exception. Since the envelope is actually dropped here, we also need to reject and drop the context.